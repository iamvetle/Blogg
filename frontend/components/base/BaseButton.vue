<template>
    <button data-test="button">{{ text }}</button>
</template>

<script setup lang="ts">

defineProps<{
    text: string
}>()

</script>

<style scoped></style>