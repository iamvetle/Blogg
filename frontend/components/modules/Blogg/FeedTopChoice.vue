<template>
	<div class="flex-row">
		<div class="grid gap-8 grid-cols-12 flex items-center">
			<div class="col-start-1 col-end-9">
				<!-- Only show the two options when web client is authenticated-->
				<div v-if="authStore.isAuthenticated" class="text-center space-x-6">
					<FeedTopChoiceFeedOptionButton
						data-test="feed_all_posts_option_button"
						:show-as-selected="showFollowingPosts"
						@show-all-feed-posts="handleShowAllFeedPosts"
					/>
					<FeedTopChoiceFollowingOptionButton
						data-test="following_posts_option"
						:show-as-selected="
							showFollowingPosts === false ? true : false
						"
						@show-following-posts="handleShowFollowingPosts"
					/>
				</div>
			</div>
			<div class="col-span-4" data-test="feed_filter_component">
				<FeedDropdownFilter v-if="showFollowingPosts === false" />
			</div>
		</div>

		<!-- IF the following buttons is pressed-->
		<div v-if="showFollowingPosts">
			<!-- Following NO one-->
			<p class="text-lg" v-if="num_of_following === 0">
				You are not following anyone.
			</p>
			<!-- Following someone but with NO posts -->
			<p
				class="text-lg"
				v-if="num_of_following && posts_by_following === 0"
			>
				None of whom you are following have published anything.
			</p>
		</div>
	</div>
</template>

<script setup lang="ts">
/**
 * Has two buttons where you can switch between seeing all posts and only the ones
 * from whom you are followning
 *
 * And the option to filter with tags
 */

const postStore = usePostStore();
const paginationStore = usePaginationStore();
const searchStore = useSearchStore();
const loggedInUserStore = useLoggedInUserStore();
const authStore = useAuthStore()

/**
 * FEED or FOLLOWING button selected display
 *
 * When the buttons are clicked this changes (also by navbar search)
 *
 * If the active url is the following url, this turns true, otherwise, it is false
 *
 * It uses the loggedinuserstore and the poststore therefore the parentcomponent makes sure those are available before rendering these
 */
const showFollowingPosts = computed(() => {
	if (paginationStore?.activeFetchURL === urls.api.posts.following) {
		return true;
	} else {
		return false;
	}
});

const num_of_following = computed(() => {
	if (
		loggedInUserStore.loggedInUserProfile &&
		loggedInUserStore.loggedInUserProfile.num_of_following > 0
	) {
		return loggedInUserStore.loggedInUserProfile?.num_of_following;
	} else {
		return 0;
	}
});

const posts_by_following = computed(() => {
	if (postStore.posts && postStore.posts.results.length > 0) {
		return postStore.posts.results.length;
	} else {
		return 0;
	}
});

/**
 * This is called when the 'feed button' is clicked.
 *
 * It changes the api endpoint url from where posts are fetched to
 * the main feed one. It then fetches all posts.
 *
 * It has its base here - the url.
 */
const handleShowAllFeedPosts = async () => {
	searchStore.resetStore();

	paginationStore.activeFetchURL = urls.api.posts.feed;
	await getPostMultipleSnippet(paginationStore.activeFetchURL);
};

/**
 * This is called when the 'feed button' is clicked.
 *
 * It changes the api endpoint url from where posts are fetched to
 * the main feed one. It then fetches all posts.
 */
const handleShowFollowingPosts = async () => {
	searchStore.resetStore();

	paginationStore.activeFetchURL = urls.api.posts.following;
	await getPostMultipleFollowingOnly(paginationStore.activeFetchURL);
};
</script>

<style scoped></style>
