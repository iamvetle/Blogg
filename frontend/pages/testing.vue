<template>
  <div class="flex">

    <form @submit.prevent method="post" class="flex flex-col" enctype="multipart/form-data">

      <label for="name">Name:</label>
      <input type="text" name="name" id="name">

      <label for="email">Email</label>
      <input type="email" name="email" id="email">

      <label for="message">Message:</label>
      <textarea name="message" id="" cols="30" rows="10" placeholder="Write a message here"></textarea>

      <label for="file">File:</label>
      <input type="file" name="file" id="file">

      <button type="submit">Submit</button>
    </form>

  </div>
</template>

<script setup lang="ts">
onMounted(() => {

  const form = document.querySelector("form")

  form?.addEventListener("submit", (e: any) => {


    const formData = new FormData(form)

    for (let item of formData) {
      console.log(item[0], item[1])
    }
    
    // ta axios of prÃ¸v fetch her
  })





})

// watchEffect(() => {

// })

</script>

<style scoped></style>