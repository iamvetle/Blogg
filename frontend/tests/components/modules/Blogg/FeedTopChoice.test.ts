import FeedTopChoice from '~/components/modules/Blogg/FeedTopChoice.vue';
import { VueWrapper, shallowMount } from '@vue/test-utils';
import { createTestingPinia } from '@pinia/testing';
import BaseButton from '~/components/base/BaseButton.vue';

let wrapper: VueWrapper;
let pinia: any = createTestingPinia();

// let generalStore; 
// let postStore; 
// let loggedInUserStore; 
// let paginationStore; 

const factory = () => {
    return shallowMount(FeedTopChoice, {
        global: {
            plugins: [pinia],
            components: {
                BaseButton
            },
            mocks: {},
            stubs: {},
        },
        props: {
            followingSelected: true
        },
        slots: {}
    })
};

describe('', () => {

    beforeEach(() => {
        // generalStore = useGeneralStore(pinia); 
        // postStore = usePostStore(pinia); 
        // loggedInUserStore = useLoggedInUserStore(pinia); 
        // paginationStore = usePaginationStore(pinia); 

    });

    afterEach(() => {
        if (wrapper) {
            wrapper.unmount();
        }
    });
    test('Should exist', () => {
        wrapper = factory()

        expect(wrapper.exists()).toBe(true)
    })
    test('Should render two (base)buttons', () => {
        const buttons = wrapper.findAllComponents({ name: "BaseButton" })

        expect(buttons).toHaveLength(2)
    })

    test('Should match snapshot', () => {
        expect(wrapper.html()).toMatchSnapshot()
    })
});