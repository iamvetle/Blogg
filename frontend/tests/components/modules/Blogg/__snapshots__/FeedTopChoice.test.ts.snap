// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Should match snapshot 1`] = `
"<div><span class=\\"flex space-x-8 justify-center\\"><base-button-stub text=\\"Feed\\" class=\\"p-2 rounded-lg bg-primary text-onPrimary border\\" data-test=\\"feed-posts-option\\"></base-button-stub><base-button-stub text=\\"Following\\" class=\\"p-2 rounded-lg bg-onPrimary text-primary border-primary border shadow-md\\" data-test=\\"following-posts-option\\"></base-button-stub></span><!-- IF the following buttons is pressed-->
  <!--v-if-->
</div>"
`;
